src/SharedMemoryParcel.php

```diff
@@ -0,0 +1,445 @@
+<?php
+
+namespace Amp\Sync;
+
+use Amp\Serialization\NativeSerializer;
+use Amp\Serialization\Serializer;
+use Amp\Sync\PosixSemaphore;
+use Amp\Sync\SyncException;
+
+/**
+ * A container object for sharing a value across contexts.
+ *
+ * A shared object is a container that stores an object inside shared memory.
+ * The object can be accessed and mutated by any thread or process. The shared
+ * object handle itself is serializable and can be sent to any thread or process
+ * to give access to the value that is shared in the container.
+ *
+ * Because each shared object uses its own shared memory segment, it is much
+ * more efficient to store a larger object containing many values inside a
+ * single shared container than to use many small shared containers.
+ *
+ * Note that accessing a shared object is not atomic. Access to a shared object
+ * should be protected with a mutex to preserve data integrity.
+ *
+ * When used with forking, the object must be created prior to forking for both
+ * processes to access the synchronized object.
+ *
+ * @see http://php.net/manual/en/book.shmop.php The shared memory extension.
+ * @see http://man7.org/linux/man-pages/man2/shmctl.2.html How shared memory works on Linux.
+ * @see https://msdn.microsoft.com/en-us/library/ms810613.aspx How shared memory works on Windows.
+ *
+ * @template T
+ * @template-implements Parcel<T>
+ */
+final class SharedMemoryParcel implements Parcel
+{
+    /** @var int The byte offset to the start of the object data in memory. */
+    private const MEM_DATA_OFFSET = 7;
+
+    // A list of valid states the object can be in.
+    private const STATE_UNALLOCATED = 0;
+    private const STATE_ALLOCATED = 1;
+    private const STATE_MOVED = 2;
+    private const STATE_FREED = 3;
+
+    /**
+     * @param string          $id
+     * @param mixed           $value
+     * @param int             $size The initial size in bytes of the shared memory segment. It will automatically be
+     *     expanded as necessary.
+     * @param int             $permissions Permissions to access the semaphore. Use file permission format specified as
+     *     0xxx.
+     * @param Serializer|null $serializer
+     *
+     * @return self
+     *
+     * @throws ParcelException
+     * @throws SyncException
+     * @throws \Error If the size or permissions are invalid.
+     */
+    public static function create(
+        string $id,
```

Comment:
The `$id` is something we likely should autogenerate, but not sure whether the API of IPC_PRIVATE is exposed.

Reply:
Since we need to open the segment in multiple processes, I don't think IPC_PRIVATE applies here.